wip-0003.md

```diff
@@ -0,0 +1,136 @@
+
+**`secp256k1bn256Map` mapping.** A new mapping to be stored in memory that is updated every time a new key is proposed. This mapping does not need to be stored on-chain, but can be constructed by information that is on-chain. This can be implemented as an additional mapping to be stored in the chain manager.
+
+``` rust
+type secp256k1bn256Map = HashMap<PublicKeyHash, Bn256PublicKey>;
+```
+
+#### Commitment Transactions
+
+**`bn256_pub_key` field.** One new field in the commitment transaction body (`CommitTransactionBody`) data structure `bn256_pub_key` is defined:
+
+```protobuf
+message CommitTransactionBody {
+    Hash dr_pointer = 1;
+    Hash commitment = 2;
+    DataRequestEligibilityClaim proof = 3;
+    repeated Input inputs = 4;
+    repeated ValueTransferOutput outputs = 5;
+    Bn256PublicKey bn256_pub_key = 6;
+}
+```
+
+#### Block Proposals
+
+**`bn256_pub_key` field.** One new field in the Block Header (`BlockHeader`) data structure `bn256_pub_key` is defined:
+
+```protobuf
+message BlockHeader {
+        message BlockMerkleRoots {
+            Hash mint_hash = 1;
+            Hash vt_hash_merkle_root = 2;
+            Hash dr_hash_merkle_root = 3;
+            Hash commit_hash_merkle_root = 4;
+            Hash reveal_hash_merkle_root = 5;
+            Hash tally_hash_merkle_root = 6;
+        }
+        uint32 version = 1;
+        CheckpointBeacon beacon = 2;
+        BlockMerkleRoots merkle_roots = 3;
+        BlockEligibilityClaim proof = 4;
+        Bn256PublicKey bn256_pub_key = 5;
+
+}
+```
+
+#### Updating the mapping
+Every time a commitment transaction associated with a pkh introduces a new bls_key, the mapping is updated. In this case we can foresee 3 scenarios:
+
+- Nodes do not change their BLS key: If nodes behave honestly, we can expect that they always specify the same bls_key in their commitment transactions.
+
+- Nodes introduce a new BLS key: There should be no problem on specifying a new bls_key as this will be updated in the mapping.
+
+- Nodes specify someone elses BLS key: We can suspect here that there is an attack going on. However such an attack would be identical to convincing a BLS key holder to vote for some specific value. Therefore, preventing this behavior does not close any attack vector.
+
+#### BLS associated ARS merkle root
+With this information, a block producer can insert a merkle root of those BLS keys in the block header. This allows a smart contract to verify a potential membership to such set.
+
+#### Backwards compatibility
+UTXOs and reputation points are still linked to a pkh, which is generated by a hash of the Secp256k1 public key. Therefore, no further change should be necessary to be compatible with previous features.
+
+#### Extendability
+Extending such functionality to be able to associate a secp256k1 public key to an additional curve key should be as simple as creating declaring an struct with all curves we want to store the public key for.
+
+``` rust
+struct BlsKeys {
```

Comment:
Why only BLS curves?
Maybe in the future we might want have other curve types.

Maybe something like:
```suggestion
struct AdditionalKeys {
```